### ISSUE TITLE
Setting algorithm to "auto" in shap.Explainer() class throws error due to missing masker

### ISSUE THREAD

**Issue by navdeep-G** [shap/shap#2460]
* The following code:

```
import pandas as pd

df = pd.read_csv(os.path.join(MLIBRARY_HOME, "tests/data/creditcard.csv"))
target_col = "default payment next month"
X,y = df.drop(target_col,axis=1), df[target_col]

gradient_booster = GradientBoostingClassifier(learning_rate=0.1)
gradient_booster.fit(X, y)

explainer= shap.Explainer(model=gradient_booster.predict, output_names=list(X.columns))

explainer(X)
```

leads to the following error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [4], in <cell line: 14>()
     10 gradient_booster.fit(X, y)
     12 explainer = shap.Explainer(model=gradient_booster.predict, output_names=list(X.columns))
---> 14 explainer(X)

File ~/Desktop/mli-2/env_38/lib/python3.8/site-packages/shap/explainers/_permutation.py:74, in Permutation.__call__(self, max_evals, main_effects, error_bounds, batch_size, outputs, silent, *args)
     70 def __call__(self, *args, max_evals=500, main_effects=False, error_bounds=False, batch_size="auto",
     71              outputs=None, silent=False):
     72     """ Explain the output of the model on the given arguments.
     73     """
---> 74     return super().__call__(
     75         *args, max_evals=max_evals, main_effects=main_effects, error_bounds=error_bounds, batch_size=batch_size,
     76         outputs=outputs, silent=silent
     77     )

File ~/Desktop/mli-2/env_38/lib/python3.8/site-packages/shap/explainers/_explainer.py:258, in Explainer.__call__(self, max_evals, main_effects, error_bounds, batch_size, outputs, silent, *args, **kwargs)
    256     feature_names = [[] for _ in range(len(args))]
    257 for row_args in show_progress(zip(*args), num_rows, self.__class__.__name__+" explainer", silent):
--> 258     row_result = self.explain_row(
    259         *row_args, max_evals=max_evals, main_effects=main_effects, error_bounds=error_bounds,
    260         batch_size=batch_size, outputs=outputs, silent=silent, **kwargs
    261     )
    262     values.append(row_result.get("values", None))
    263     output_indices.append(row_result.get("output_indices", None))

File ~/Desktop/mli-2/env_38/lib/python3.8/site-packages/shap/explainers/_permutation.py:132, in Permutation.explain_row(self, max_evals, main_effects, error_bounds, batch_size, outputs, silent, *row_args)
    129     i += 1
    131 # evaluate the masked model
--> 132 outputs = fm(masks, zero_index=0, batch_size=batch_size)
    134 if row_values is None:
    135     row_values = np.zeros((len(fm),) + outputs.shape[1:])

File ~/Desktop/mli-2/env_38/lib/python3.8/site-packages/shap/utils/_masked_model.py:64, in MaskedModel.__call__(self, masks, zero_index, batch_size)
     62         full_masks = np.zeros((int(np.sum(masks >= 0)), self._masker_cols), dtype=np.bool)
     63         _convert_delta_mask_to_full(masks, full_masks)
---> 64         return self._full_masking_call(full_masks, zero_index=zero_index, batch_size=batch_size)
     66 else:
     67     return self._full_masking_call(masks, batch_size=batch_size)

File ~/Desktop/mli-2/env_38/lib/python3.8/site-packages/shap/utils/_masked_model.py:93, in MaskedModel._full_masking_call(self, masks, zero_index, batch_size)
     91     masked_inputs = self.masker(delta_ind, *self.args).copy()
     92 else:
---> 93     masked_inputs = self.masker(mask, *self.args)
     95 # wrap the masked inputs if they are not already in a tuple
     96 if not isinstance(masked_inputs, tuple):

TypeError: 'NoneType' object is not callable
```

* I believe this is due to how the `shap` handles arbitrary models that are callable, which seems to set the `algorithm` argument to `permutation`, which requires a masker.

* I have attached the dataset for ease of repro.
[creditcard.csv](https://github.com/slundberg/shap/files/8354338/creditcard.csv)

* Note, I understand I can just pass in the gradient booster model. However, this is the easiest way to show my use case and it also eases reproducibility.

**Comment by 20032303092** [comment]
do you solve it, sir?

**Comment by ascripter** [comment]
The problem still persists in shap 0.42

**Comment by thatlittleboy** [comment]
I believe this is the same issue raised in #3310.

The solution is to pass in a masker during the init explicitly. To make this more apparent I included a check and raised an error during the initialisation, PR #3315. Hopefully this is a better user experience.
It should already be included in the latest release 0.43.

I'll close this issue as duplicate for now
